### PR TITLE
refactor: use `usize` for `Array`, `ArrayBuffer`, `TypedArray`, `DataView`

### LINKED ISSUE #4766: refactor `DataView`, `TypedArray`, `ArrayBuffer` indices to use `usize`
The usage of `u64` vs `usize` is not unified for indices of the types backed by `ArrayBuffer`.
We want to use `usize` wherever possible.

For example:
https://github.com/boa-dev/boa/blob/27959b2cebc00bbd917e8a8f3782b0b6b27a3bfb/core/engine/src/object/builtins/jsdataview.rs#L180-L186

https://github.com/boa-dev/boa/blob/27959b2cebc00bbd917e8a8f3782b0b6b27a3bfb/core/engine/src/builtins/array_buffer/mod.rs#L200-L213

### PR DESCRIPTION
**Pull request by zhuzhu81998**
Closes #4766 .

Many lines of changes are just deleting `as usize` really.

### CODE REVIEW

File: core/engine/src/builtins/array/from_async.rs [hunk 1]
```diff
@@ -248,7 +248,7 @@ fn from_async_iterator(
                     iterator_record,
                 } => {
                     // Inverted conditional makes for a simpler code.
-                    if k < 2u64.pow(53) - 1 {
+                    if (k as f64) < Number::MAX_SAFE_INTEGER {
```
Comment (hansl) [inline core/engine/src/builtins/array/from_async.rs]: No need to convert to a float.

Comment (zhuzhu81998) [reply core/engine/src/builtins/array/from_async.rs]: I assume you mean the I64 here https://github.com/zhuzhu81998/boa/blob/22e6a146961845cd5d2d89db8860d52c2e0c0e57/core/engine/src/value/conversions/try_into_js.rs#L76-L77? That one is not public right now. For now I will cast both to `u64` to ensure comparison makes sense. In a separate PR we could introduce `Numer::MAX_SAFE_INTEGER_U64` (and get rid of some `as u64` here and elsewhere). What do you think?

**Comment by zhuzhu81998** [comment]
just saw #4083 :sweat_smile:  (@jedel1043 ig the "spec-related checks" reasons are rediscovered)
however the theoretically possible case of the length of e.g. a string being over `usize` will never happen in practice. both firefox and chrome has the length limit well below a `usize` on 32-bit: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length .

**Comment by jedel1043** [comment]
I think we just need to ensure we evaluate spec-related checks before rejecting indices bigger than `usize` (probably)

**Comment by jedel1043** [comment]
For good measure we should still test this on a 32-bit system

**Comment by zhuzhu81998** [comment]
> I think we just need to ensure we evaluate spec-related checks before rejecting indices bigger than `usize` (probably)

well if we dont reject indices bigger than `usize` at the latest the OS would anyway?

**Comment by zhuzhu81998** [comment]
> For good measure we should still test this on a 32-bit system

hmm dont have one. dont think github action runner has that either

**Comment by zhuzhu81998** [comment]
> hmm dont have one. dont think github action runner has that either

cross-compiling does not work out of the box because of `openssl-sys` :roll_eyes: .
But even without this PR, tests on 32-bit target fail:

```
test test::weak::miri::eph_self_referential_chain ... FAILED
test test::weak::miri::eph_self_referential ... FAILED
test test::allocation::miri::gc_recursion ... ok

failures:

---- test::weak::miri::eph_self_referential_chain stdout ----

thread '<unnamed>' (373283) panicked at core/gc/src/test/mod.rs:42:13:
assertion `left == right` failed
  left: 108
 right: 168
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/std/src/panicking.rs:689:5
   1: core::panicking::panic_fmt
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/panicking.rs:80:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/panicking.rs:394:5
   4: boa_gc::test::Harness::assert_exact_bytes_allocated::{{closure}}
             at ./src/test/mod.rs:42:13
   5: std::thread::local::LocalKey<T>::try_with
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:513:12
   6: std::thread::local::LocalKey<T>::with
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:477:20
   7: boa_gc::test::Harness::assert_exact_bytes_allocated
             at ./src/test/mod.rs:40:16
   8: boa_gc::test::weak::miri::eph_self_referential_chain::{{closure}}
             at ./src/test/weak.rs:215:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

thread 'test::weak::miri::eph_self_referential_chain' (373277) panicked at core/gc/src/test/weak.rs:186:9:
called `Result::unwrap()` on an `Err` value: Any { .. }
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/std/src/panicking.rs:689:5
   1: core::panicking::panic_fmt
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/panicking.rs:80:14
   2: core::result::unwrap_failed
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/result.rs:1867:5
   3: core::result::Result<T,E>::unwrap
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1233:23
   4: boa_gc::test::run_test
             at ./src/test/mod.rs:50:19
   5: boa_gc::test::weak::miri::eph_self_referential_chain
             at ./src/test/weak.rs:186:9
   6: boa_gc::test::weak::miri::eph_self_referential_chain::{{closure}}
             at ./src/test/weak.rs:181:36
   7: core::ops::function::FnOnce::call_once
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- test::weak::miri::eph_self_referential stdout ----

thread '<unnamed>' (373275) panicked at core/gc/src/test/mod.rs:42:13:
assertion `left == right` failed
  left: 36
 right: 56
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/std/src/panicking.rs:689:5
   1: core::panicking::panic_fmt
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/panicking.rs:80:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/panicking.rs:394:5
   4: boa_gc::test::Harness::assert_exact_bytes_allocated::{{closure}}
             at ./src/test/mod.rs:42:13
   5: std::thread::local::LocalKey<T>::try_with
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:513:12
   6: std::thread::local::LocalKey<T>::with
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:477:20
   7: boa_gc::test::Harness::assert_exact_bytes_allocated
             at ./src/test/mod.rs:40:16
   8: boa_gc::test::weak::miri::eph_self_referential::{{closure}}
             at ./src/test/weak.rs:169:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

thread 'test::weak::miri::eph_self_referential' (373274) panicked at core/gc/src/test/weak.rs:153:9:
called `Result::unwrap()` on an `Err` value: Any { .. }
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/std/src/panicking.rs:689:5
   1: core::panicking::panic_fmt
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/panicking.rs:80:14
   2: core::result::unwrap_failed
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/result.rs:1867:5
   3: core::result::Result<T,E>::unwrap
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1233:23
   4: boa_gc::test::run_test
             at ./src/test/mod.rs:50:19
   5: boa_gc::test::weak::miri::eph_self_referential
             at ./src/test/weak.rs:153:9
   6: boa_gc::test::weak::miri::eph_self_referential::{{closure}}
             at ./src/test/weak.rs:144:30
   7: core::ops::function::FnOnce::call_once
             at /usr/local/rustup/toolchains/1.93.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    test::weak::miri::eph_self_referential
    test::weak::miri::eph_self_referential_chain

test result: FAILED. 19 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.16s

error: test failed, to rerun pass `-p boa_gc --lib`
```

**Comment by jedel1043** [comment]
Oof, let me see if I can fix those

**Comment by zhuzhu81998** [comment]
> Oof, let me see if I can fix those

I think the problem is that on 32-bit targets (or anything that is not 64-bit), the exact bytes allocated will be different, because of the different pointer sizes, so this will fail :joy: :
https://github.com/boa-dev/boa/blob/94360bfb3471b7bb433358c2c2d2e1d4dabd5e3c/core/gc/src/test/weak.rs#L215-L215

**Comment by jedel1043** [comment]
#5090 added CI for 32-bit targets, so it should be much easier to catch bugs on this now

**Comment by zhuzhu81998** [comment]
> #5090 added CI for 32-bit targets, so it should be much easier to catch bugs on this now

very nice.

**Comment by hansl** [comment]
I can also test this on my ARMv7, or you can use QEMU to spin up an emulated one.

**Comment by jedel1043** [comment]
Ran the test262 suite using cross on the `i686-unknown-linux-gnu` target and I got this list of failing tests:

Broken tests (55):
test/RegExp/prototype/Symbol.replace/coerce-lastindex.js (previously Passed)
test/RegExp/prototype/exec/failure-lastindex-set.js (previously Passed)
test/ArrayBuffer/data-allocation-after-object-creation.js (previously Passed)
test/Array/prototype/includes/length-boundaries.js (previously Passed)
test/Array/prototype/unshift/length-near-integer-limit.js (previously Passed)
test/Array/prototype/unshift/clamps-to-integer-limit.js (previously Passed)
test/Array/prototype/unshift/throws-if-integer-limit-exceeded.js (previously Passed)
test/Array/prototype/copyWithin/length-near-integer-limit.js (previously Passed)
test/Array/prototype/push/S15.4.4.7_A4_T2.js (previously Passed)
test/Array/prototype/push/S15.4.4.7_A4_T1.js (previously Passed)
test/Array/prototype/push/length-near-integer-limit.js (previously Passed)
test/Array/prototype/push/S15.4.4.7_A3.js (previously Passed)
test/Array/prototype/push/S15.4.4.7_A2_T2.js (previously Passed)
test/Array/prototype/push/clamps-to-integer-limit.js (previously Passed)
test/Array/prototype/push/throws-if-integer-limit-exceeded.js (previously Passed)
test/Array/prototype/push/length-near-integer-limit-set-failure.js (previously Passed)
test/Array/prototype/fill/length-near-integer-limit.js (previously Passed)
test/Array/prototype/concat/arg-length-exceeding-integer-limit.js (previously Passed)
test/Array/prototype/concat/arg-length-near-integer-limit.js (previously Passed)
test/Array/prototype/indexOf/15.4.4.14-3-28.js (previously Passed)
test/Array/prototype/indexOf/length-near-integer-limit.js (previously Passed)
test/Array/prototype/indexOf/15.4.4.14-3-8.js (previously Passed)
test/Array/prototype/indexOf/15.4.4.14-3-14.js (previously Passed)
test/Array/prototype/indexOf/15.4.4.14-3-29.js (previously Passed)
test/Array/prototype/splice/length-exceeding-integer-limit-shrink-array.js (previously Passed)
test/Array/prototype/splice/length-and-deleteCount-exceeding-integer-limit.js (previously Passed)
test/Array/prototype/splice/create-species-length-exceeding-integer-limit.js (previously Passed)
test/Array/prototype/splice/length-near-integer-limit-grow-array.js (previously Passed)
test/Array/prototype/splice/clamps-length-to-integer-limit.js (previously Passed)
test/Array/prototype/splice/throws-if-integer-limit-exceeded.js (previously Passed)
test/Array/prototype/splice/S15.4.4.12_A3_T1.js (previously Passed)
test/Array/prototype/reduceRight/length-near-integer-limit.js (previously Passed)
test/Array/prototype/findLast/maximum-index.js (previously Passed)
test/Array/prototype/every/15.4.4.16-3-14.js (previously Passed)
test/Array/prototype/every/15.4.4.16-3-29.js (previously Passed)
test/Array/prototype/every/15.4.4.16-3-8.js (previously Passed)
test/Array/prototype/findLastIndex/maximum-index.js (previously Passed)
test/Array/prototype/pop/S15.4.4.6_A3_T2.js (previously Passed)
test/Array/prototype/pop/length-near-integer-limit.js (previously Passed)
test/Array/prototype/pop/S15.4.4.6_A2_T2.js (previously Passed)
test/Array/prototype/pop/S15.4.4.6_A3_T1.js (previously Passed)
test/Array/prototype/pop/clamps-to-integer-limit.js (previously Passed)
test/Array/prototype/reverse/length-exceeding-integer-limit-with-proxy.js (previously Passed)
test/Array/prototype/reverse/length-exceeding-integer-limit-with-object.js (previously Passed)
test/Array/prototype/some/15.4.4.17-3-8.js (previously Passed)
test/Array/prototype/some/15.4.4.17-3-29.js (previously Passed)
test/Array/prototype/some/15.4.4.17-3-14.js (previously Passed)
test/Array/prototype/some/15.4.4.17-3-28.js (previously Passed)
test/Array/prototype/lastIndexOf/length-near-integer-limit.js (previously Passed)
test/Array/prototype/lastIndexOf/15.4.4.15-3-28.js (previously Passed)
test/Array/prototype/toSpliced/length-clamped-to-2pow53minus1.js (previously Passed)
test/Array/prototype/toSpliced/length-exceeding-array-length-limit.js (previously Passed)
test/Array/prototype/slice/length-exceeding-integer-limit-proxied-array.js (previously Passed)
test/Array/prototype/slice/length-exceeding-integer-limit.js (previously Passed)
test/SharedArrayBuffer/data-allocation-after-object-creation.js (previously Passed)

**Comment by zhuzhu81998** [comment]
thanks, will have a look later.

**Comment by jedel1043** [comment]
Btw if you install `cross` it is kinda easy to run individual tests to see if they work on 32-bit machines:
```
cross run --target i686-unknown-linux-gnu --bin boa_tester -- run -s test/Array/prototype/some/15.4.4.17-3-28.js
```
Just don't run the _full_ suite because it consistently OOM'd for me, too much memory usage from all the thousands of tests running haha

**Comment by zhuzhu81998** [comment]
```
`/workspaces/boa/test262/test/built-ins/Array/prototype/concat/arg-length-exceeding-integer-limit.js`: starting
`/workspaces/boa/test262/test/built-ins/Array/prototype/concat/arg-length-exceeding-integer-limit.js`: Failed
`/workspaces/boa/test262/test/built-ins/Array/prototype/concat/arg-length-exceeding-integer-limit.js`: result text
Uncaught Test262Error {
    message: "[1].concat(spreadableLengthOutOfRange) throws a TypeError exception Expected a TypeError but got a RangeError"
}
    at <anonymous> (/workspaces/boa/test262/harness/assert.js:95:13)
    at <main> (/workspaces/boa/test262/test/built-ins/Array/prototype/concat/arg-length-exceeding-integer-limit.js:26:14)
```

sigh....

**Comment by zhuzhu81998** [comment]
I had to take another look at the spec because of these failing tests.

The problem seems to be that since "array" can also just be a normal object (in which case index is technically speaking not index, but rather a normal `PropertyKey` which could be an index if it is inside `u32` or it would be a string instead), thus it is expected that an engine accepts something as an "index" even though the platform can not technically handle it....

I wish I never touched this... But it seems that the main branch implementation is, spec-wise, strictly speaking, largely correct with the using `u64` everywhere and "unsafely" converting it to `usize`. The test suite does not test for the platform-specific edge cases and the spec says nothing about implementation details anyway.

There are 1 or 2 small issues with e.g. `byte_offset` which returns a `usize` though. We should probably just use `u64` as much as possible. Type casting like might not be free, but should be efficient enough such that it does not matter.

Btw, technically spec says "suitable", but does not consider platform at all (not even a little bit).
image

So unless there is something else, I think we can close this PR, and I will open another one which adjusts fewer things to ensure the `u64` principle.
